compare258.c

```diff
@@ -0,0 +1,204 @@
+/* compare258.c -- aligned and unaligned versions of compare258
+ * Copyright (C) 2020 Nathan Moinvaziri
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+#include "zbuild.h"
+#include "zutil.h"
+
+#include "fallback_builtins.h"
+
+/* ALIGNED, byte comparison */
+static inline int32_t compare256_c_static(const unsigned char *src0, const unsigned char *src1) {
+    const unsigned char *src0start = src0;
+    const unsigned char *src0end = src0 + 256;
+
+    do {
+        if (*src0 != *src1)
+            break;
+        src0 += 1, src1 += 1;
+        if (*src0 != *src1)
+            break;
+        src0 += 1, src1 += 1;
+        if (*src0 != *src1)
+            break;
+        src0 += 1, src1 += 1;
+        if (*src0 != *src1)
+            break;
+        src0 += 1, src1 += 1;
+    } while (src0 < src0end);
+
+    return (int32_t)(src0 - src0start);
+}
+
+static inline int32_t compare258_c_static(const unsigned char *src0, const unsigned char *src1) {
+    if (*src0 != *src1)
+        return 1;
+    src0 += 1, src1 += 1;
+    if (*src0 != *src1)
+        return 2;
+    src0 += 1, src1 += 1;
+    if (*src0 != *src1)
+        return 3;
```

Comment:
This makes no sense at all... 258 is multiple of many numbers and 258-compare can be implemented more efficiently without calling 256-byte compare.

Reply:
It depends on `MIN_MATCH`. Longest_match will use COMPARE256 if MIN_MATCH > 1 and COMPARE258 if MIN_MATCH <= 1

Reply:
It is 256 because the initial match check is at least 2 bytes (uint16_t)

Reply:
Hmm, is MIN_MATCH lower than 3 even possible/allowed/sensible? I thought 3 was the minimum, but that you could increase this to avoid having short matches.

Reply:
258 is multiple of 6, 256 is multiple of 4 and 8... This means it is more efficient to use 6 checks per iteration when comparing 258 bytes, but 4 checks per iteration when comparing 256 bytes.

Only few processors have enough general-purpose registers to do 8 checks per iteration efficiently.

Reply:
@Dead2 It is just to make sure the initial 2 bytes match before going to the optimized version that compares 256 bytes instead of 258 bytes... It's inefficient to use intrinsics for the 2 bytes.

Reply:
Yes, I just don't understand @nmoinvaz comment about MIN_MATCH <= 1. That doesn't make sense to me.

Reply:
@Dead2 To me neither... But that's why we ask clarifications... Some times we write something we don't mean to due to distractions ;)